delta/plugins/composite-views/src/main/scala/ch/epfl/bluebrain/nexus/delta/plugins/compositeviews/CompositeViewFactory.scala

```diff
@@ -0,0 +1,143 @@
+package ch.epfl.bluebrain.nexus.delta.plugins.compositeviews
+
+import cats.data.NonEmptyList
+import ch.epfl.bluebrain.nexus.delta.kernel.utils.UUIDF
+import ch.epfl.bluebrain.nexus.delta.plugins.compositeviews.model.CompositeView.RebuildStrategy
+import ch.epfl.bluebrain.nexus.delta.plugins.compositeviews.model.{CompositeViewFields, CompositeViewProjection, CompositeViewProjectionFields, CompositeViewSource, CompositeViewSourceFields, CompositeViewValue}
+import ch.epfl.bluebrain.nexus.delta.rdf.IriOrBNode.Iri
+import ch.epfl.bluebrain.nexus.delta.sdk.projects.model.ProjectBase
+import monix.bio.UIO
+
+object CompositeViewFactory {
+
+  /**
+    * Create a new [[CompositeViewValue]] from [[CompositeViewFields]].
+    *
+    *   - The indexing revision for the common spaces and the projections defaults to 1.
+    */
+  def create(fields: CompositeViewFields)(implicit projectBase: ProjectBase, uuidF: UUIDF): UIO[CompositeViewValue] =
+    for {
+      sources     <- fields.sources.traverse { create }
+      projections <- fields.projections.traverse { create(_, 1) }
+    } yield CompositeViewValue(
+      fields.name,
+      fields.description,
+      1,
+      sources.toNem,
+      projections.toNem,
+      fields.rebuildStrategy
+    )
+
+  /**
+    * Update a [[CompositeViewValue]] from [[CompositeViewFields]] and the previous [[CompositeViewValue]].
+    *
+    *   - When a source or a projection keep the same id, the existing uuid is also preserved
+    *   - When any source is added/updated/deleted, the indexing revision for the common space and all the projections
+    *     are updated
+    *   - When a projection is updated, only the indexing revision of this projection is updated
+    */
+  def update(fields: CompositeViewFields, current: CompositeViewValue, nextRev: Int)(implicit
+      projectBase: ProjectBase,
+      uuidF: UUIDF
+  ): UIO[CompositeViewValue] = {
+    for {
+      sources             <- fields.sources.traverse { upsert(_, current.sources.lookup) }.map(_.toNem)
+      // If any source has changed, we update the indexing rev for sources
+      sourceHasChanged     = current.sources != sources
+      newSourceIndexingRev = if (sourceHasChanged) nextRev else current.sourceIndexingRev
```

Comment:
If the source does not change, but the projection does, the CompositeViewValue does not have a new indexing rev. But then we are in fact reindexing but the name of the projection does change? Maybe I am just confused. But this is how I understand the failing supervision test